lib/ClangImporter/ClangImporter.cpp

```diff
@@ -7471,6 +7471,37 @@ bool importer::isForeignReferenceTypeWithoutImmortalAttrs(const clang::QualType
          !hasImmortalAtts(pointeeType->getDecl());
 }
 
+// TODO: Instead of just checking the immediate parent, do a traversal till the
+// top most parent
+std::pair<bool, clang::CXXRecordDecl *>
+getFRTInfoFromBaseDecls(swift::ASTContext &ctx, Evaluator &evaluator,
+                        const clang::RecordDecl *decl) {
+  bool declHasAReferenceTypeBase = false;
+  clang::CXXRecordDecl *baseRefDecl = nullptr;
+
+  if (auto *cxxRecordDecl = llvm::dyn_cast<clang::CXXRecordDecl>(decl)) {
+    auto hasReferenceTypeBase =
+        [&evaluator, &ctx, &baseRefDecl](const clang::CXXBaseSpecifier &base) {
+          if (clang::CXXRecordDecl *baseDecl =
+                  base.getType()->getAsCXXRecordDecl()) {
+            auto semanticsKind = evaluateOrDefault(
+                evaluator, CxxRecordSemantics({baseDecl, ctx}), {});
+            if (semanticsKind == CxxRecordSemanticsKind::Reference) {
```

Comment:
I am not 100% sure if it is enough to only check the `CxxRecordSemanticsKind`. One question is how we want to handle `SWIFT_IMMORTAL_REFERENCE` cases. E.g., if a class has a `SWIFT_IMMORTAL_REFERENCE` base and a `SWIFT_SHARED_REFERENCE` base is that an error or is it OK? Does having a `SWIFT_IMMORTAL_REFERENCE` base implies that the derived classes should also be `SWIFT_IMMORTAL_REFERENCE`s?

Should the user have an attribute to overwrite this default behavior and force importing a type as a value type?

Reply:
> E.g., if a class has a `SWIFT_IMMORTAL_REFERENCE` base and a `SWIFT_SHARED_REFERENCE` base is that an error or is it OK?

I think for now it would be best to make this an error. In the future we can consider allowing this hierarchy and assuming `SWIFT_SHARED_REFERENCE` semantics for the derived class.

> Does having a `SWIFT_IMMORTAL_REFERENCE` base implies that the derived classes should also be `SWIFT_IMMORTAL_REFERENCE`s?

I think that's reasonable, yes.